dirty_cat/similarity_encoder.py

```diff
@@ -30,58 +34,68 @@
+        i: int,
+        ngram_range: Tuple[int, int]) -> None:
+    """
+    Update inplace a dict of similarities between a string and a vocabulary
 
     Parameters
     ----------
     x_count_vector: np.array
-        count vector of the sample based on the ngrams of the vocabulary
+        Count vector of the sample based on the ngrams of the vocabulary
     vocabulary_count_matrix: np.array
-        count vector of the vocabulary based on its ngrams
+        Count vector of the vocabulary based on its ngrams
     str_x: str
-        the actual sample string
+        The actual sample string
     vocabulary_ngram_counts: np.array
-        number of ngrams for each unique element of the vocabulary
+        Number of ngrams for each unique element of the vocabulary
     se_dict: dict
-        dictionary containing the similarities for each x in unq_X
+        Dictionary containing the similarities for each x in unq_X
     unq_X: np.array
-        the arrayes of all unique samples
+        The arrays of all unique samples
     i: str
-        the index of x_count_vector in the csr count matrix
+        The index of x_count_vector in the csr count matrix
     ngram_range: tuple
-
+        n-grams to use for the decomposition, where `n_min <= n <= n_max`.
     """
     nonzero_idx = x_count_vector.indices
     nonzero_vals = x_count_vector.data
 
-    samegrams = np.asarray(
+    same_grams = np.asarray(
         (vocabulary_count_matrix[:, nonzero_idx].minimum(nonzero_vals)).sum(
             axis=1))
 
-    allgrams = get_ngram_count(
-        str_x, ngram_range) + vocabulary_ngram_counts - samegrams
-    similarity = np.divide(samegrams, allgrams, out=np.zeros_like(samegrams),
-                           where=allgrams != 0)
+    all_grams = get_ngram_count(
+        str_x, ngram_range) + vocabulary_ngram_counts - same_grams
+    similarity = np.divide(same_grams, all_grams,
+                           out=np.zeros_like(same_grams),
+                           where=all_grams != 0)
     se_dict[unq_X[i]] = similarity.reshape(-1)
 
 
-def ngram_similarity(X, cats, ngram_range, hashing_dim, dtype=np.float64):
+def ngram_similarity(X,
+                     cats: List[str],
+                     ngram_range: Tuple[int, int],
+                     hashing_dim: int,
+                     dtype: type = np.float64) -> np.array:
     """
     Similarity encoding for dirty categorical variables:
-        Given to arrays of strings, returns the
-        similarity encoding matrix of size
-        len(X) x len(cats)
+        Given two arrays of strings, returns the similarity encoding matrix
+        of size len(X) x len(cats)
 
     ngram_sim(s_i, s_j) =
         ||min(ci, cj)||_1 / (||ci||_1 + ||cj||_1 - ||min(ci, cj)||_1)
     """
     min_n, max_n = ngram_range
     unq_X = np.unique(X)
-    cats = np.array([' %s ' % cat for cat in cats])
-    unq_X_ = np.array([' %s ' % x for x in unq_X])
+    cats = np.array([f' {cat} ' for cat in cats])
```

Comment:
Did you time the difference (with large arrays)? I worry that one might be faster than the other.

Reply:
Indeed, there is a difference as highlighted [here](https://colab.research.google.com/drive/1rfQTSHBDm3aH9u4C4WkZG2p9oyCXQU3g?usp=sharing) and [here](https://stackoverflow.com/questions/10530315/str-performance-in-python).